src/api.ts

```diff
@@ -0,0 +1,249 @@
+import { EventEmitter } from 'eventemitter3'
+import { AMCPCommand, Commands } from './commands'
+import { Connection, ResponseTypes } from './connection'
+
+export interface Options {
+	/** Host name of the machine to connect to. Defaults to 127.0.0.1 */
+	host?: string
+	/** Port number to connect to. Defaults to 5250 */
+	port?: number
+	/** If true, the library will not use request id's and instead wait for the command to return before sending the next */
+	useSequential?: boolean
+	/** Minimum amount of time before a request is considered to be timed out */
+	timeoutTime?: number
+}
+
+export interface SendResult {
+	sentOk: boolean
+	error?: Error
+}
+
+export interface Request {
+	/**
+	 * The generated request ID
+	 */
+	requestId?: string
+	/**
+	 * Resolves when the request is sent to CasparCG, result is true if it was sent successfully
+	 */
+	sent: Promise<SendResult>
+	/**
+	 * Added to the object after the sent promise has resolved, this resolves when CasparCG has executed the request
+	 */
+	response?: Promise<Response>
+	/**
+	 * Removes the request from the request queue
+	 */
+	cancel: () => void
+}
+
+interface InternalRequest {
+	requestId?: string
+	command: AMCPCommand
+	request: Request
+
+	resolve: (response: Response) => void
+	reject: (error: Error) => void
+
+	processed: boolean
+	processedTime?: number
+	sentResolve: (sent: SendResult) => void
+	// sentReject: (error: Error) => void
+}
+
+export interface Response {
+	reqId?: string
+	command: Commands
+	responseCode: number
+	data: any[]
+
+	type: ResponseTypes
+	message: string
+}
+
+export type ConnectionEvents = {
+	connect: []
+	disconnect: []
+	error: [error: Error]
+}
+
+export class BasicCasparCGAPI extends EventEmitter<ConnectionEvents> {
+	private _connection: Connection
+	private _host: string
+	private _port: number
+
+	private readonly _useSequential: boolean
+	private _requestQueue: Array<InternalRequest> = []
+	private _timeoutTimer: NodeJS.Timer
+	private _timeoutTime: number
+
+	constructor(options?: Options) {
+		super()
+
+		this._host = options?.host || '127.0.0.1'
+		this._port = options?.port || 5250
+		this._useSequential = options?.useSequential || false
+
+		this._connection = new Connection(this._host, this._port)
+
+		this._connection.on('connect', () => {
+			this.emit('connect')
+			this._processQueue().catch((e) => this.emit('error', e))
+		})
+		this._connection.on('disconnect', () => this.emit('disconnect'))
+
+		this._connection.on('data', (response) => {
+			if (this._useSequential) {
+				const request = this._requestQueue.shift()
+
+				if (request) {
+					request.resolve(response)
+				}
+			} else {
+				const request = this._requestQueue.find((req) => req.requestId === response.reqId)
+
+				if (request) {
+					request.resolve(response)
+					this._requestQueue = this._requestQueue.filter((req) => req.requestId !== response.reqId)
+				}
+			}
+
+			this._processQueue().catch(() => null)
```

Comment:
Swallowing errors like this can be dangerous. You should at least emit the error if there is one